### ISSUE TITLE
Is the use of catch_unwind in remote_handle correct?

### ISSUE THREAD

**Issue by najamelan** [rust-lang/futures-rs#2077]
@Nemo157 Would you mind clarifying the reasoning behind the use of `catch_unwind` here?

Im looking into [this code](https://github.com/rust-lang/futures-rs/blob/771513cea996bac0c8c4017be687e289c966f504/futures-util/src/future/future/remote_handle.rs#L100-L114):

```rust
pub(super) fn remote_handle<Fut: Future>(future: Fut) -> (Remote<Fut>, RemoteHandle<Fut::Output>) {
    let (tx, rx) = oneshot::channel();
    let keep_running = Arc::new(AtomicBool::new(false));

    // AssertUnwindSafe is used here because `Send + 'static` is basically
    // an alias for an implementation of the `UnwindSafe` trait but we can't
    // express that in the standard library right now.
    let wrapped = Remote {
        future: AssertUnwindSafe(future).catch_unwind(),
        tx: Some(tx),
        keep_running: keep_running.clone(),
    };

    (wrapped, RemoteHandle { rx, keep_running })
}
```

I can see a few potential problems:
1. The comment suggests that `Send + 'static` is `UnwindSafe`, however we don't require `Send + 'static` bounds on the future we assert unwind safe.
2. `Send + 'static != UnwindSafe`. There has been a lot of discussion on this around the [RFC](https://github.com/rust-lang/rfcs/blob/master/text/1236-stabilize-catch-panic.md) for `catch_unwind`. Thus it seems that even if the future were `Send + 'static`, we assert that it is `UnwindSafe` when it actually might not be, which means users still need to be aware of unwind safety (not an easy concept), and the documentation doesn't mention any of this.
3. So far, apart from here, the only place I have been able to find `catch_unwind` is in `tokio::task::harness`. It seems executors from the futures lib, async-std, wasm-bindgen-futures all don't catch panics. This makes me wonder, why here?
4. There is a `Send + 'static` bound on the `Future::Output` of the future. I wonder why it is there, since even though right now `remote_handle` can be used for `!Send` futures, it still requires the output to be `Send`, seemingly artificially limiting it's use cases. I reason that if a panic happens in the future, the output will not be produced and thus cannot be observed by other code, so it probably has no effect on unwind safety. I'm working on a lib that introduces a new `JoinHandle` type that wraps the one from tokio, the one from async-std and that I want to also work with the executors from the futures lib, in order to enable more executor agnostic code. However if I were to roll a custom `remote_handle` to overcome this restriction, I really need to understand the intricacies of unwind safety first, hence this question. 

For the record, the use of `AssertUnwindSafe` in tokio also seems incorrect to me, but at least they require `Send + 'static` on the future (which whacks some of the moles but not all).

**Comment by Nemo157** [comment]
`AssertUnwindSafe` and `catch_unwind` are safe APIs, therefore any use of them is _sound_. Whether it is _correct_, I'm not sure, I personally don't strongly understand the meaning of `AssertUnwindSafe`. That comment has been moved through a few refactorings from https://github.com/rust-lang/futures-rs/commit/732dabb16504ddb58858e0a5a113e82bd948274d#diff-7921ee9b6e376f6df138b33d31d95975R54-R56, maybe @cramertj can explain.

In regards to other executors not catching panics, that is an issue I have seen with them all. For example with `futures::executor::ThreadPool` panicking in tasks will slowly deplete the pool of workers and end up deadlocking the application when it runs out. In `tokio` (0.1) I tried setting up an application to log then cleanly exit when a panic occurred, but couldn't work out any way to structure it so that that would occur without triggering panic-during-unwind in `tokio` internals. I think there's an ecosystem wide issue with things not being resilient to running without `panic=abort` that should be addressed at some point.

The reason `RemoteHandle` specifically catches panics is the same reason `std::thread::spawn` does, it allows propagating the panic back to whatever is waiting on the result, instead of `RemoteHandle.await`/`ThreadHandle.join()` just infinitely blocking.

**Comment by najamelan** [comment]
@nemo, sorry my use of "sound" is probably wrong here. If we are in doubt, I think @alexcrichton has most expertise on unwind safety.

> The reason RemoteHandle specifically catches panics is the same reason std::thread::spawn does, it allows propagating the panic back to whatever is waiting on the result, instead of RemoteHandle.await/ThreadHandle.join() just infinitely blocking.

Ok, I didn't realize it would just block. That changes things a bit. I will give it some more though. I wrote the following before your answer came through, so it might have to be revised:

If my understanding above is right, I can see a few solutions:
1. We don't catch the panic. We remove the `Send` bound on the output. This would open the use cases and leave it up to users to deal with the panic. They can always use `FutureExt::catch_unwind` if they want that.
2. If we want to keep the `catch_unwind` we can add a `Fut: UnwindSafe` bound rather than using `AssertUnwindSafe`. Add some guidance in the docs for people running into their futures not being unwind safe, because there will be a fair few.
3. If 2 is considered to harsh for ergonomics, we should at least document unwind safety issues of using this. Potentially we could require the future to be `Send + 'static` to whack some moles and kill some use cases, but I'm not very fond of this solution for the same reasons that it was decided not to have `catch_unwind` require `Send + 'static` (note that was before the `UnwindSafe` trait existed.

**Comment by najamelan** [comment]
> That comment has been moved through a few refactorings from [732dabb#diff-7921ee9b6e376f6df138b33d31d95975R54-R56](https://github.com/rust-lang/futures-rs/commit/732dabb16504ddb58858e0a5a113e82bd948274d#diff-7921ee9b6e376f6df138b33d31d95975R54-R56), maybe @cramertj can explain.

I see that in that commit the bounds on the future did include `Send + 'static` which must have been lost in translation.

**Comment by najamelan** [comment]
> I personally don't strongly understand the meaning of AssertUnwindSafe

From what I have been able to figure out after 2 days of research it boils down to this:
- when panicking, some data might be in inconsistent state, because the function returns early.
- when this inconsistent data get's observed later on, this can cause logic errors. Without unsafe code, it shouldn't cause memory unsafety, since there are no uninitialized vars in rust. The rfs lists some other mitigations which make it hard in rust to actually run into real trouble here, but it's possible.
- Observing inconsistent data happens in destructors being called while unwinding and through types that allow shared mutability because as the program will keep running, they might continue to observe the data that was manipulated by the code that panicked. Ways to do this are through anything that holds an `UnsafeCell`, like `RefCell`. TLS also is a problem, although I haven't figured out exactly the story around TLS. 
- You can do it with `Send + 'static` for example `parking_lot::Mutex` is `Send + 'static` but will not poison, and thus is `!UnwindSafe`. 

`UnwindSafe` is an auto trait which will not be implemented for anything that has `UnsafeCell` and I think `&mut` like a lot of futures might. `AssertUnwindSafe` is a bit like `impl UnwindSafe for MyType`, basically saying to the compiler "I know, this has types that are not unwind safe, but I manually checked the code in this type and nothing can go wrong. It's fine to call catch_unwind on it, I promise." That's why it's kind of dodgy to do that on other peoples types, since we basically (manually) checked shit.

I'm definitely not an expert, so take this explanation with a grain of salt, but that's what I understood.

**Comment by najamelan** [comment]
I did a little experimenting. Since the unwinding will call the destructor for the `oneshot::Sender`, the calling thread will not hang, but we do lose the error message of the original panic. It would seem we just use `catch_panic` here to capture that error message and resume the panic on the thread awaiting the handle. AFAICT, this means we don't claim it's fine for the program/thread to keep running, and we don't make any new risks concerning unwind safety than there already are in all rust code (destructors). 

I presume then the `AssertUnwindSafe` is fine. What I would propose is I file a PR that clarifies the reasoning here in comments to make it easier for future readers/maintainers to understand and that I remove the arbitrary `Send` bound on the output. Maybe that requires adding `unsafe impl<T: Send> Send for RemoteHandle<T>`. I can file 2 separate PR's, which is probably better for review.

Does that sound like a good solution?

**Comment by najamelan** [comment]
I have second thoughts. We actually catch the panic on the thread that polls the `Remote` but `resume_unwind` on the thread that awaits the `RemoteHandle`. That means that the thread on which the panic happens actually does keep running. That might be an argument for not doing the `catch_unwind` after all. If we don't catch, in the console you get:
```
thread '<unnamed>' panicked at 'DIE, thread!', src/main.rs:36:9 
thread 'main' panicked at 'oneshot canceled', <::std::macros::panic macros>:5:6
```
The first line is the actual error from the original panic. The second line is from the thread that polls `RemoteHandle`. The difference in the error message not being propagated to the thread that awaits `RemoteHandle` would only impact users if they call `catch_unwind` again and try to capture the error to do something with it (It's still a `Box<dyn Any + Send>`, so I suppose it could be downcast). Of course we can change the error message from "oneshot cancelled" to something along the lines of: "Remote future panicked".

It seems to me that currently we are tearing down the wrong thread. We can tear down both if we think that's the right option, but it's not what Rust usually does. I think choices to be made are:
- which threads should be torn down
- if we don't want to tear down the thread on which the `RemoteHandle` is awaited, do we change the API to return a `Result`. (breaking change).
- if we don't require that the error message travels to the thread awaiting `RemoteHandle`, shall we just remove `catch_unwind` all together?
- alternatively, if we don't want to remove `catch_unwind`, to forward the error, shall we tear down the thread that panicked at least?

If people give their opinion, I'm willing to file a PR.

**Comment by alexcrichton** [comment]
Ah sorry it's been a very long time for me and I don't know much about the API here. Unwind safety is pretty tricky in general but it's also not really all that strict in general. Making everything "best effort" is generally good enough from what I've seen.

**Comment by najamelan** [comment]
@alexcrichton thanks for chipping in. Maybe I can resume because the thread is getting long. Currently we take a random user supplied future we know nothing about, declare it unwind safe and let the thread on which it panics continue. As a bonus we tear down an unrelated thread. 

Does that sound like a good best effort? In the PR I propose we at least tear down the thread in which the panic happens, and leave other threads alone.

**Comment by alexcrichton** [comment]
In my personal opinion that's fine, yeah. Panic safety is often based around "you caught the panic and things were left in bad state you poked at later", but the intention here is that the "thread" of computation (the future) is just split across a few locations and we transfer the panic manually between those locations as an implementation detail. In that sense you can't easily poke at bad state because when your "thread" (future) panics it just keeps panicking.

**Comment by najamelan** [comment]
> In that sense you can't easily poke at bad state because when your "thread" (future) panics it just keeps panicking.

I'm not quite sure I understand this sentence entirely. I can image a situation where I create a future, send the handle of to another thread to be awaited and run the future locally. In that case it might well share data with other stuff in the thread like TLS or RefCells. 

Now the thread that receives the handle will be torn down but the one in which the future runs won't. It just seems backwards to me. 

Of course you could send the future to a different thread with say a `parking_lot::mutex` inside which might well corrupt data observed from the thread holding the handle, but it seems that would require a bit more effort.

**Comment by najamelan** [comment]
I suppose that the current situation does imply that if you send it on a threadpool, the pool doesn't lose it's thread, which might well be desired.
